## `nixpkgs-review` result

Generated using [`nixpkgs-review-gha`](https://github.com/Defelo/nixpkgs-review-gha)

Command: `nixpkgs-review pr 516292`
Commit: [`c30bc2b99508c664800276ba85acc90f6d2c857d`](https://github.com/NixOS/nixpkgs/commit/c30bc2b99508c664800276ba85acc90f6d2c857d) ([subsequent changes](https://github.com/NixOS/nixpkgs/compare/c30bc2b99508c664800276ba85acc90f6d2c857d..pull/516292/head))
Merge: [`a3597f1260e4cd58e6369a5a32ce08ef797bab35`](https://github.com/NixOS/nixpkgs/commit/a3597f1260e4cd58e6369a5a32ce08ef797bab35)

Logs: https://github.com/kilyanni/nixpkgs-review-gha/actions/runs/25296154891


---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>python313Packages.python-statemachine</li>
    <li>python313Packages.python-statemachine.dist</li>
    <li>python314Packages.python-statemachine</li>
    <li>python314Packages.python-statemachine.dist</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>python313Packages.python-statemachine</li>
    <li>python313Packages.python-statemachine.dist</li>
    <li>python314Packages.python-statemachine</li>
    <li>python314Packages.python-statemachine.dist</li>
  </ul>
</details>

---
### `x86_64-darwin` (sandbox = relaxed)
<details>
  <summary>:x: 4 packages failed to build:</summary>
  <ul>
    <li>python313Packages.python-statemachine</li>
    <li>python313Packages.python-statemachine.dist</li>
    <li>python314Packages.python-statemachine</li>
    <li>python314Packages.python-statemachine.dist</li>
  </ul>
</details>

---
<details>
<summary>Error logs: `x86_64-darwin`</summary>
<details>
<summary>python313Packages.python-statemachine</summary>
<pre>DEBUG    statemachine.engines.async_:async_.py:432 [SM] macrostep 1: event=__initial__
DEBUG    statemachine.engines.base:base.py:572 [SM] States to enter: {Loading}
DEBUG    statemachine.engines.async_:async_.py:237 [SM] Entering state: Loading
DEBUG    statemachine.engines.async_:async_.py:381 [SM] Macrostep 1: eventless/internal queue
DEBUG    statemachine.invoke:invoke.py:434 invoke spawn async: loading.0944aeae on state loading
DEBUG    statemachine.engines.async_:async_.py:415 [SM] Macrostep 1: external queue
DEBUG    statemachine.engines.async_:async_.py:497 [SM] Processing loop ended
DEBUG    statemachine.engines.async_:async_.py:369 [SM] Processing loop started: loading
DEBUG    statemachine.engines.async_:async_.py:381 [SM] Macrostep 1: eventless/internal queue
DEBUG    statemachine.engines.async_:async_.py:415 [SM] Macrostep 1: external queue
DEBUG    statemachine.engines.async_:async_.py:497 [SM] Processing loop ended
---------------------------- Captured log teardown -----------------------------
DEBUG    statemachine.invoke:invoke.py:472 invoke loading.0944aeae: completed (cancelled=False)
=========================== short test summary info ============================
FAILED tests/test_contrib_timeout.py::TestTimeoutBasic::test_timeout_fires_done_invoke[async] - AssertionError: assert &#x27;done&#x27; in OrderedSet([&#x27;waiting&#x27;])
FAILED tests/test_contrib_timeout.py::TestTimeoutCustomEvent::test_custom_event_fires[async] - AssertionError: assert &#x27;timed_out&#x27; in OrderedSet([&#x27;waiting&#x27;])
FAILED tests/test_invoke.py::TestInvokeGroup::test_group_returns_ordered_results[async] - AssertionError: assert &#x27;ready&#x27; in OrderedSet([&#x27;loading&#x27;])
FAILED tests/test_invoke.py::TestInvokeGroup::test_group_error_cancels_remaining[sync] - AssertionError: assert &#x27;error_state&#x27; in OrderedSet([&#x27;loading&#x27;])
FAILED tests/test_invoke.py::TestInvokeGroup::test_group_error_cancels_remaining[async] - AssertionError: assert &#x27;error_state&#x27; in OrderedSet([&#x27;loading&#x27;])
===== 5 failed, 936 passed, 28 skipped, 1 deselected, 2 xfailed in 45.81s ======</pre>
</details>
<details>
<summary>python314Packages.python-statemachine</summary>
<pre>DEBUG    statemachine.engines.async_:async_.py:415 [SM] Macrostep 0: external queue
DEBUG    statemachine.engines.async_:async_.py:432 [SM] macrostep 1: event=__initial__
DEBUG    statemachine.engines.base:base.py:572 [SM] States to enter: {Loading}
DEBUG    statemachine.engines.async_:async_.py:237 [SM] Entering state: Loading
DEBUG    statemachine.engines.async_:async_.py:381 [SM] Macrostep 1: eventless/internal queue
DEBUG    statemachine.invoke:invoke.py:434 invoke spawn async: loading.9f4837df on state loading
DEBUG    statemachine.engines.async_:async_.py:415 [SM] Macrostep 1: external queue
DEBUG    statemachine.engines.async_:async_.py:497 [SM] Processing loop ended
DEBUG    statemachine.engines.async_:async_.py:369 [SM] Processing loop started: loading
DEBUG    statemachine.engines.async_:async_.py:381 [SM] Macrostep 1: eventless/internal queue
DEBUG    statemachine.engines.async_:async_.py:415 [SM] Macrostep 1: external queue
DEBUG    statemachine.engines.async_:async_.py:497 [SM] Processing loop ended
---------------------------- Captured log teardown -----------------------------
DEBUG    statemachine.invoke:invoke.py:472 invoke loading.9f4837df: completed (cancelled=False)
=========================== short test summary info ============================
FAILED tests/test_contrib_timeout.py::TestTimeoutBasic::test_timeout_fires_done_invoke[sync] - AssertionError: assert &#x27;done&#x27; in OrderedSet([&#x27;waiting&#x27;])
FAILED tests/test_contrib_timeout.py::TestTimeoutBasic::test_timeout_fires_done_invoke[async] - AssertionError: assert &#x27;done&#x27; in OrderedSet([&#x27;waiting&#x27;])
FAILED tests/test_contrib_timeout.py::TestTimeoutCustomEvent::test_custom_event_fires[async] - AssertionError: assert &#x27;timed_out&#x27; in OrderedSet([&#x27;waiting&#x27;])
FAILED tests/test_invoke.py::TestInvokeGroup::test_group_returns_ordered_results[async] - AssertionError: assert &#x27;ready&#x27; in OrderedSet([&#x27;loading&#x27;])
===== 4 failed, 937 passed, 28 skipped, 1 deselected, 2 xfailed in 44.84s ======</pre>
</details>
</details>

---
### `aarch64-darwin` (sandbox = relaxed)
<details>
  <summary>:x: 4 packages failed to build:</summary>
  <ul>
    <li>python313Packages.python-statemachine</li>
    <li>python313Packages.python-statemachine.dist</li>
    <li>python314Packages.python-statemachine</li>
    <li>python314Packages.python-statemachine.dist</li>
  </ul>
</details>

---
<details>
<summary>Error logs: `aarch64-darwin`</summary>
<details>
<summary>python313Packages.python-statemachine</summary>
<pre>DEBUG    statemachine.engines.async_:async_.py:415 [SM] Macrostep 0: external queue
DEBUG    statemachine.engines.async_:async_.py:432 [SM] macrostep 1: event=__initial__
DEBUG    statemachine.engines.base:base.py:572 [SM] States to enter: {Loading}
DEBUG    statemachine.engines.async_:async_.py:237 [SM] Entering state: Loading
DEBUG    statemachine.engines.async_:async_.py:381 [SM] Macrostep 1: eventless/internal queue
DEBUG    statemachine.invoke:invoke.py:434 invoke spawn async: loading.2d397c78 on state loading
DEBUG    statemachine.engines.async_:async_.py:415 [SM] Macrostep 1: external queue
DEBUG    statemachine.engines.async_:async_.py:497 [SM] Processing loop ended
DEBUG    statemachine.engines.async_:async_.py:369 [SM] Processing loop started: loading
DEBUG    statemachine.engines.async_:async_.py:381 [SM] Macrostep 1: eventless/internal queue
DEBUG    statemachine.engines.async_:async_.py:415 [SM] Macrostep 1: external queue
DEBUG    statemachine.engines.async_:async_.py:497 [SM] Processing loop ended
---------------------------- Captured log teardown -----------------------------
DEBUG    statemachine.invoke:invoke.py:472 invoke loading.2d397c78: completed (cancelled=False)
=========================== short test summary info ============================
FAILED tests/test_contrib_timeout.py::TestTimeoutComposition::test_timeout_fires_before_slow_invoke[sync] - AssertionError: assert &#x27;stuck&#x27; in OrderedSet([&#x27;loading&#x27;])
FAILED tests/test_contrib_timeout.py::TestTimeoutComposition::test_timeout_fires_before_slow_invoke[async] - AssertionError: assert &#x27;stuck&#x27; in OrderedSet([&#x27;loading&#x27;])
FAILED tests/test_invoke.py::TestInvokeGroup::test_group_returns_ordered_results[async] - AssertionError: assert &#x27;ready&#x27; in OrderedSet([&#x27;loading&#x27;])
FAILED tests/test_invoke.py::TestInvokeGroup::test_group_error_cancels_remaining[async] - AssertionError: assert &#x27;error_state&#x27; in OrderedSet([&#x27;loading&#x27;])
===== 4 failed, 937 passed, 28 skipped, 1 deselected, 2 xfailed in 29.98s ======</pre>
</details>
<details>
<summary>python314Packages.python-statemachine</summary>
<pre>
tests/test_invoke.py:418: AssertionError
------------------------------ Captured log setup ------------------------------
DEBUG    asyncio:selector_events.py:64 Using selector: KqueueSelector
------------------------------ Captured log call -------------------------------
DEBUG    statemachine.engines.base:base.py:572 [SM] States to enter: {Loading}
DEBUG    statemachine.engines.base:base.py:668 [SM] Entering state: Loading
DEBUG    statemachine.engines.sync:sync.py:80 [SM] Processing loop started: loading
DEBUG    statemachine.engines.sync:sync.py:95 [SM] Macrostep 0: eventless/internal queue
DEBUG    statemachine.invoke:invoke.py:365 invoke spawn sync: loading.18ecf180 on state loading
DEBUG    statemachine.engines.sync:sync.py:133 [SM] Macrostep 0: external queue
DEBUG    statemachine.engines.sync:sync.py:180 [SM] Processing loop ended
DEBUG    statemachine.engines.sync:sync.py:80 [SM] Processing loop started: loading
DEBUG    statemachine.engines.sync:sync.py:95 [SM] Macrostep 0: eventless/internal queue
DEBUG    statemachine.engines.sync:sync.py:133 [SM] Macrostep 0: external queue
DEBUG    statemachine.engines.sync:sync.py:180 [SM] Processing loop ended
=========================== short test summary info ============================
FAILED tests/test_contrib_timeout.py::TestTimeoutComposition::test_timeout_fires_before_slow_invoke[sync] - AssertionError: assert &#x27;stuck&#x27; in OrderedSet([&#x27;loading&#x27;])
FAILED tests/test_invoke.py::TestInvokeGroup::test_group_error_cancels_remaining[sync] - AssertionError: assert &#x27;error_state&#x27; in OrderedSet([&#x27;loading&#x27;])
===== 2 failed, 939 passed, 28 skipped, 1 deselected, 2 xfailed in 30.05s ======</pre>
</details>
</details>

--- a/pkgs/development/python-modules/python-statemachine/default.nix
+++ b/pkgs/development/python-modules/python-statemachine/default.nix
@@ -1,0 +1,73 @@
+{
+  lib,
+  stdenv,
+  buildPythonPackage,
+  fetchPypi,
+  hatchling,
+  pytestCheckHook,
+  pytest-asyncio,
+  pytest-benchmark,
+  pytest-mock,
+  pytest-timeout,
+  pytest-django,
+  pydot,
+}:
+
+buildPythonPackage (finalAttrs: {
+  pname = "python-statemachine";
+  version = "3.0.0";
+  pyproject = true;
+
+  __structuredAttrs = true;
+
+  src = fetchPypi {
+    pname = "python_statemachine";
+    inherit (finalAttrs) version;
+    hash = "sha256-kVI/nq1zwdb+zJddXG4L/jY/v1N8XwvzCbzQ+U+UQbI=";
+  };
+
+  build-system = [ hatchling ];
+
+  nativeCheckInputs = [
+    pytestCheckHook
+    pytest-asyncio
+    pytest-benchmark
+    pytest-mock
+    pytest-timeout
+    pytest-django
+    pydot
+  ];
+
+  pytestFlags = [ "--benchmark-disable" ];
+
+  disabledTestPaths = [
+    # quite slow
+    "tests/test_weighted_transitions.py"
+    "tests/scxml/"
+  ];
+
+  disabledTests = [
+    # broken upstream
+    "statemachine.contrib.diagram.quickchart_write_svg"
+  ]
+  ++ lib.optionals stdenv.hostPlatform.isDarwin [
+    # async state transitions appear broken on darwin
+    # TODO: investigate on an actual macOS machine
+    "test_timeout_fires_done_invoke"
+    "test_timeout_fires_before_slow_invoke"
+    "test_custom_event_fires"
+    "test_group_returns_ordered_results"
+    "test_group_error_cancels_remaining"
+  ];
+
+  pythonImportsCheck = [ "statemachine" ];
+
+  meta = {
+    description = "Expressive statecharts and FSMs for modern Python";
+
+    homepage = "https://github.com/fgmacedo/python-statemachine";
+    changelog = "https://github.com/fgmacedo/python-statemachine/releases/tag/v${finalAttrs.version}";
+    license = lib.licenses.mit;
+    maintainers = with lib.maintainers; [ kilyanni ];
+  };
+})

--- a/pkgs/development/python-modules/python-statemachine/default.nix
+++ b/pkgs/development/python-modules/python-statemachine/default.nix
@@ -1,0 +1,79 @@
+{
+  lib,
+  stdenv,
+  buildPythonPackage,
+  fetchPypi,
+  hatchling,
+  pytestCheckHook,
+  pytest-asyncio,
+  pytest-benchmark,
+  pytest-mock,
+  pytest-timeout,
+  pytest-django,
+  pydot,
+}:
+
+buildPythonPackage (finalAttrs: {
+  pname = "python-statemachine";
+  version = "3.0.0";
+  pyproject = true;
+
+  __structuredAttrs = true;
+
+  src = fetchPypi {
+    pname = "python_statemachine";
+    inherit (finalAttrs) version;
+    hash = "sha256-kVI/nq1zwdb+zJddXG4L/jY/v1N8XwvzCbzQ+U+UQbI=";
+  };
+
+  build-system = [ hatchling ];
+
+  nativeCheckInputs = [
+    pytestCheckHook
+    pytest-asyncio
+    pytest-benchmark
+    pytest-mock
+    pytest-timeout
+    pytest-django
+    pydot
+  ];
+
+  pytestFlags = [ "--benchmark-disable" ];
+
+  disabledTestPaths = [
+    # quite slow
+    "tests/test_weighted_transitions.py"
+    "tests/scxml/"
+  ];
+
+  disabledTests = [
+    # broken upstream
+    "statemachine.contrib.diagram.quickchart_write_svg"
+  ]
+  ++ lib.optionals stdenv.hostPlatform.isDarwin [
+    # async state transitions appear broken on darwin
+    # TODO: investigate on an actual macOS machine
+    "test_timeout_fires_done_invoke"
+    "test_timeout_fires_before_slow_invoke"
+    "test_custom_event_fires"
+
+    # flaky
+    "test_group_returns_ordered_results"
+    "test_group_error_cancels_remaining"
+    "test_group_with_file_io"
+    "test_group_cancel_on_exit"
+    "test_group_single_callable"
+    "test_each_sm_instance_gets_own_group"
+  ];
+
+  pythonImportsCheck = [ "statemachine" ];
+
+  meta = {
+    description = "Expressive statecharts and FSMs for modern Python";
+
+    homepage = "https://github.com/fgmacedo/python-statemachine";
+    changelog = "https://github.com/fgmacedo/python-statemachine/releases/tag/v${finalAttrs.version}";
+    license = lib.licenses.mit;
+    maintainers = with lib.maintainers; [ kilyanni ];
+  };
+})

--- a/pkgs/top-level/python-packages.nix
+++ b/pkgs/top-level/python-packages.nix
@@ -16005,6 +16005,8 @@ self: super: with self; {
 
   python-sql = callPackage ../development/python-modules/python-sql { };
 
+  python-statemachine = callPackage ../development/python-modules/python-statemachine { };
+
   python-status = callPackage ../development/python-modules/python-status { };
 
   python-stdnum = callPackage ../development/python-modules/python-stdnum { };

